### PR TITLE
added jax.scipy.special.spence function 

### PR DESCRIPTION
**Pull request by lucashofer**
I added jax.scipy.special.spence to JAX. The implementation below is based on the Cephes library algorithm which SciPy uses, but written in JAX code. The author of the Cephes library has already given permission with the JAX license.

### CODE REVIEW

**Comment by jakevdp** [comment]
Hi - thanks for the PR. If this is adapted from code from another package, the implementation should go in the third_party directory, along with an appropriate license statement: https://github.com/google/jax/tree/main/jax/_src/third_party/scipy. Thanks!

**Comment by lucashofer** [comment]
Hi Jake,

The approval for using the Cephes software algorithms is given in 

[https://github.com/google/jax/blob/main/jax/_src/scipy/special.py](https://github.com/google/jax/blob/main/jax/_src/scipy/special.py) 

as an email exchange with the Cephes author (line 225). Since a couple other functions in that file also use Cephes algorithms (the ntdr and ndtri functions) would you still like me to move the spence function to the third party directory?

If so, I can create a separate subdirectory for Cephes algorithms and copy the email exchange in for the LICENSE.txt file.

**Comment by jakevdp** [comment]
Thanks - I think we're covered then. But please modify the comment here: https://github.com/google/jax/blob/8268cd562de9b2cb1dc5693ee3e84619d42ad12e/jax/_src/scipy/special.py#L225

And mention spence and the URL from which the values were used, if different from ndtr and ndtri

File: tests/lax_scipy_test.py [hunk 1]
```diff
@@ -621,6 +621,34 @@ def test_vq(self, n_obs, n_codes, n_feats, dtype):
     self._CheckAgainstNumpy(osp_cluster.vq.vq, lsp_cluster.vq.vq, args_maker, check_dtypes=False)
     self._CompileAndCheck(lsp_cluster.vq.vq, args_maker)
 
+  def test_spence(self):
+    """Check against known values of Spence's function. SciPy implements
+    this test with the same values internally, but tests them against the
+    known analytic values. We test against SciPy values."""
```
Comment (jakevdp) [inline tests/lax_scipy_test.py:627]: Rather than testing only against a 1D array of specific values, it would be better do do a parameterized test over values of different shapes and dtypes, just to make sure we're covering all corner cases.

File: tests/lax_scipy_test.py [hunk 2]
```diff
@@ -621,6 +621,34 @@ def test_vq(self, n_obs, n_codes, n_feats, dtype):
...
+      """Check against known values of Spence's function. SciPy implements
+      this test with the same values internally, but tests them against the
+      known analytic values. We test against SciPy values instead."""
+      dataset = [1.0, 2.0, 0.5, 0.0, (-1 + 5 ** .5) / 2, (3 - 5 ** .5) / 2,
+                (1 + 5 ** .5) / 2,(3 + 5 ** .5) / 2]
+      x = np.array([dataset])
+      return x
+
+    with self.subTest('Test JIT Against Numpy for Analytically Known Values'):
+          self._CheckAgainstNumpy(osp_special.spence, lsp_special.spence, args_maker,
```
Comment (jakevdp) [inline tests/lax_scipy_test.py:639]: Fix indentation please

File: jax/_src/scipy/special.py [hunk 3]
```diff
@@ -1518,3 +1521,88 @@ def exp1(x: ArrayLike, module='scipy.special') -> Array:
...
+    \begin{equation}
+    -\int_0^z \frac{\log(1 - t)}{t}dt
+    \end{equation}
+  this is our spence(1 - z).
+  """
+  x = jnp.asarray(x)
+  dtype = lax.dtype(x)
+  if dtype not in (jnp.float32, jnp.float64):
+    raise TypeError(
+        "x.dtype={} is not supported, see docstring for supported types."
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:1606]: Use an f-string

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks good! A few comments below. Also:

- please add the function to `jax.scipy.rst` so it will appear in the documentation
- after addressing the comments, please squash your changes into a single commit (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests)

Thanks!

**Comment by lucashofer** [comment]
Hi Jake,

I fixed all the comments above. However, I was having trouble with the squashing and rebasing locally. 

I ended up creating a new branch locally with all the changes and then was able to successfully rebase it after fetching changes from the main jax repo. Force pushing it seems to have created some errors. I'll keep working on this.

**Comment by lucashofer** [comment]
I've submitted a new PR with all the changes and none of the conflicts at:

[#14408](https://github.com/google/jax/pull/14408/commits)

Again, sorry for the hassle.

**Comment by jakevdp** [comment]
Closing in favor of #14408
